src/main/java/learnositysdk/request/Init.java

```diff
@@ -156,7 +155,72 @@ public void setAction(String action) throws Exception
     	// Re-generate the signature, as an action is now set
         this.securityPacket.put("signature", this.generateSignature());
     }
-    
+
+    /**
+     * Generate the data necessary to make a request to one of the
+     * Learnosity products/services - note this doesn't escape any of the json
+     * strings so remains valid json.
+     * @return  The signed JSON Object ready to be used as a request to learnosity by the front-end
+     */
+    public JSONObject generateAsJsonObject() throws Exception
```

Comment:
This method seems to duplicate the behaviour of `generate`. I wonder if you could use one or the other as the reference, and do a conversion from/to JSON.

I think it would actually make more sense to keep your new method as the core, as it generates a more usable object. The `generate` method could then simply encode the returned `JSONObject` and return as a JSON `String`.

Reply:
@shtrom @jjwatson the problem with returning the request as a JSON Object is that the order of keys in JSON objects is no fixed. Some languages (e.g. PHP) keep the order, but Java doesn't. So if the request is returned as an object here, and then it's stringified and sent to the backend, the order of keys might change and the signatures won't match anymore. Possibly this can be fixed by adding the request and any other details which are used in signature generation as string. Passing them as objects is definitely a problem.

Reply:
@klauste - I see, thank you

I'll happily take the string provided it's unescaped, as it's the escaping that's the problem. I'll add a new PR request that simply returns the string unescaped.

Reply:
So, to summarise my understanding, we need to generate the signature on a JSON String that has been escaped in some way. Before creating this string, we do create a `JsonObject` internally. Perhaps it would be best to store that object, and possibly encode the request `String` at the last minute.

Reply:
@shtrom @jjwatson I'm not sure which problem we're trying to solve, so here's some background on the code, hopefully it's helpful to do whatever needs to be done: the type of requestPacket which you can pass to the Init constructor is Object https://github.com/Learnosity/learnosity-sdk-java/blob/master/src/main/java/learnositysdk/request/Init.java#L133. This allows a user to pass in any Object which can be converted into a JSON Object, e.g. a Map, a String or a JSON Object. If a String is passed, it is not changed before signature generation, but used as is (https://github.com/Learnosity/learnosity-sdk-java/blob/master/src/main/java/learnositysdk/request/Init.java#L383). If it's not a string, it is parsed to a string and then some escaping is done: https://github.com/Learnosity/learnosity-sdk-java/blob/master/src/main/java/learnositysdk/request/Init.java#L396 After that, this.requestString is never transformed back to an object again, it is used for signature generation as is and will be passed to author api which sends it unchanged to our backend. As long as string encoding is the same, this ensures that the signature generation in the backend will be the same as in the sdk.

this.requestPacket is a JSON object generated from the argument in the Init constructor. It's only used in method setServiceOptions to get some values. Otherwise it plays no part in signature generation and is also not used in the generate method.

So if there are some issues in escaping, they could be fixed in the validateRequestPacket method. I would not recommend returning the request as an object, as this will most probably cause issues with signature generation in the backend.

Reply:
The issue is line 402 in Init. Json and java share many escape characters (for example newline being \n). Unescaping them if they are present causes the json to be invalid. This isn't a problem in the Items API where these characters would be unlikely to appear, but in the questions API for example - the template field can easily have escaped new lines, which when unescaped causes the json to be messed up and unusable. I'm not sure why they're being unescaped, but that was the issue I was trying to work around. Sorry my communication was a little off.

This unescaping doesn't occur if a string is sent in though - so I just ensure I toString my JSONObject before I send it in to Init(), meaning there's a perfectly usable workaround.